xs/src/libslic3r/IO/TMF.hpp

```diff
@@ -0,0 +1,246 @@
+            {"relationships", "http://schemas.openxmlformats.org/package/2006/relationships"} // Relationships namespace.
+    };
+    ///< Namespaces in the 3MF document.
+
+    enum material_groups_types{
+        UNKNOWEN,
+        BASE_MATERIAL,
+        COLOR,
+        TEXTURE,
+        COMPOSITE_MATERIAL,
+        MULTI_PROPERTIES
+    };
+    ///< 3MF material groups in 3MF core and in 3MF materials extension.
+
+    TMFEditor(std::string input_file, Model* model);
+
+    /// Write TMF function called by TMF::write() function.
+    bool produce_TMF();
+
+    /// Read TMF function called by TMF::read() function.
+    bool consume_TMF();
+
+    ~TMFEditor();
+
+private:
+    std::string zip_name; ///< The zip archive file name.
+    Model* model; ///< The model to be read or written.
+    zip_t* zip_archive; ///< The zip archive object for reading/writing zip files.
+    std::string buff; ///< The buffer currently used in write functions.
+    ///< When it reaches a max capacity it's written to the current entry in the zip file.
+    int material_group_id; ///< The id available for the next material property group to be written.
+    int object_id; ///< The id available for the next object to be written.
+    std::map<int, int> material_groups_ids; ///<
+
+    /// Write the necessary types in the 3MF package. This function is called by produceTMF() function.
+    bool write_types();
+
+    /// Write the necessary relationships in the 3MF package. This function is called by produceTMF() function.
+    bool write_relationships();
+
+    /// Write the Model in a zip file. This function is called by produceTMF() function.
+    bool write_model();
+
+    /// Write the metadata of the model. This function is called by writeModel() function.
+    bool write_metadata();
+
+    /// Write Materials. The current supported materials are only of the core specifications.
+    // The 3MF core specs support base materials, Each has by default color and name attributes.
+    bool write_materials();
+
+    /// Write object of the current model. This function is called by writeModel() function.
+    /// \param index int the index of the object to be read
+    /// \return bool 1: write operation is successful , otherwise not.
+    bool write_object(int index);
+
+    /// Write the build element.
+    bool write_build();
+
+    /// Read the Model.
+    bool read_model();
+
+    // Helper Functions.
+    /// Append the buffer with a string to be written. This function calls write_buffer() if the buffer reached its capacity.
+    void append_buffer(std::string s);
+
+    /// This function writes the buffer to the current opened zip entry file if it exceeds a certain capacity.
+    void write_buffer();
+
+    template <class T>
+    std::string to_string(T number){
+        std::ostringstream s;
+        s << number;
+        return s.str();
+    }
+
+};
+
+/// 3MF XML Document parser.
+struct TMFParserContext{
+
+    enum TMFNodeType {
```

Comment:
C++11 note: use scoped enums where possible (enum class); this will scope them appropriately. 

Suggest pushing it up outside the struct (as it's public already); this is not a hard requirement (and subject to additional comment by @alexrj )